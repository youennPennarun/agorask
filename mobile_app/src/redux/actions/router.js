/* @flow */
export const PUSH = 'ROUTER_PUSH';
export const POP = 'ROUTER_POP';
export const TRANSITION_START = 'ROUTER_TRANSITION_START';
export const TRANSITION_END = 'ROUTER_TRANSITION_END';


export function login() {
  return pushRoute({key: 'login'});
}

export function pushRoute(route: Object = {}) {
  const routeConfig = {
    ...route,
    routeKey: route.key
  };
  return {
    type: PUSH,
    route: routeConfig,
  };
}

export function popRoute() {
  return {
    type: POP,
  };
}

export function transitionStart() {
  return {
    type: TRANSITION_START,
  };
}

export function transitionEnd() {
  return {
    type: TRANSITION_END,
  };
}

