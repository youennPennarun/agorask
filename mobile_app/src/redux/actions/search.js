/* @flow */
/* global fetch */
import Config from 'react-native-config';

export const UPDATE_QUERY = 'UPDATE_QUERY_SEARCH_VENUES';
export const CLEAR = 'CLEAR_SEARCH_VENUES';
export const FETCHING_LIST = 'FETCHING_LIST_SEARCH_VENUES';
export const SUCCESS_LIST = 'SUCCESS_LIST_SEARCH_VENUES';
export const FAILURE_LIST = 'FAILURE_LISTE_SEARCH_VENUES';

export function updateQuery(query) {
  return {
    type: UPDATE_QUERY,
    query,
  };
}

export function clear() {
  return {
    type: CLEAR,
  };
}

export function searchVenue(radius = 3000) {
  return (dispatch, getState) => {
    const {query} = getState().search;
    const {coords} = getState().userLocation;

    const queryParams = `?lat=${coords.lat}&lng=${coords.lng}&query=${query}&radius=${radius}`;
    dispatch({
      type: FETCHING_LIST,
    });
    fetch(`${Config.API_URL}/venues/search${queryParams}`, {
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
      },
    })
    .then((response): Promise<*> => {
      console.log(response);
      response.json();
    })
    .then((venues: Array) => {
      console.log("=========================");
      console.log(venues);
      console.log("=========================");
      dispatch({
        type: SUCCESS_LIST,
        venues,
      });
    })
    .catch(error => {
      dispatch({
        type: FAILURE_LIST,
        error,
      });
    });
  };
}
