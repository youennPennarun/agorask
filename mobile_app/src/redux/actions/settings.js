/* @flow */
import TaskChecker from '../../components/natives/TaskChecker';

export const SET_NOTIFICATIONS = 'SETTINGS_SET_NOTIFICATIONS';

export function setNotifications(enabled: Boolean): {type: string, enabled: Boolean} {
  if (enabled) {
    TaskChecker.start();
  } else {
    TaskChecker.stop();
  }
  return {
    type: SET_NOTIFICATIONS,
    enabled,
  };
}
