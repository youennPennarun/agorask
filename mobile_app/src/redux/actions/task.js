/* @flow */
/* global fetch */
import {API_URL} from 'react-native-config';

export const FETCHING_TASK = 'FETCH_TASK';
export const SUCCESS_FETCH_TASK = 'SUCCESS_FETCH_TASK';

const TASKS_API_URL = `${API_URL}/tasks`;

export function getTask(id): Function {
  return dispatch => {
    dispatch({type: FETCHING_TASK, id});

    fetch(`${TASKS_API_URL}/${id}`)
      .then((response): Promise => {
        if (response.status === 200) {
          return response.json();
        }
        throw new Error(`Server responded with status ${response.status}`);
      })
      .then(task => { dispatch({type: SUCCESS_FETCH_TASK, task, id}); })
      .catch(e => { console.error(e); });
  };
}

