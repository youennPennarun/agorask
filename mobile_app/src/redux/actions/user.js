/* @flow */
import {AsyncStorage} from 'react-native';
import Config from 'react-native-config';

export const LOGIN = 'USER_LOGIN';
export const LOGIN_SUCCESS = 'USER_LOGIN_SUCCESS';
export const LOGIN_FAIL = 'USER_LOGIN_FAIL';
export const DISCONNECT = 'USER_DISCONNECT';

export const LOGOUT = 'USER_LOGIN';

export const UPDATE_LOCATION = 'UPDATE_USER_LOCATION';

const config = {
  API_URL: Config.API_URL,
};

export function updateUserLocation(lat: Number, lng: Number): Object {
  return {
    type: UPDATE_LOCATION,
    lat,
    lng,
  };
}

function loginRequested(): {type: string} {
  return {
    type: LOGIN,
  };
}

function success(username: string, token: string): {type: string, token: string} {
  return {
    type: LOGIN_SUCCESS,
    token,
    username,
  };
}
function failed(message: string): {type: string, message: string} {
  return {
    type: LOGIN_FAIL,
    message,
  };
}

export function loadTokenFromStorage(): Function {
  return dispatch => {
    AsyncStorage.getItem('agorask:user')
      .then(userStr => {
        let username;
        let token;
        try {
          const parsed = JSON.parse(userStr);
          username = parsed.username;
          token = parsed.token;
        } catch (e) {}
        dispatch(success(username, token));
      })
      .catch(console.err);
  };
}

export function disconnect(): Function {
  return dispatch => {
    AsyncStorage.removeItem('agorask:user')
      .then(() => {
        dispatch({
          type: DISCONNECT,
        });
      }).catch(console.err);
  };
}


export function login(username: string, password: string): Function {
  return (dispatch: Function): Promise => {
    dispatch(loginRequested());
    return fetch(`${config.API_URL}/users/login`, {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        username,
        password,
      }),
    })
    .then((response: Object): Promise<Object> => {
      if (response.status === 200) {
        return response.json();
      } else if (response.status === 401) {
        return Promise.reject(new Error('Invalid credentials'));
      }
      return Promise.reject(new Error(response.text));
    })
    .then(({token}) => {
      dispatch(success(username, token));
      return AsyncStorage.setItem('agorask:user', JSON.stringify({username, token}));
    })
    .then(() => {
    })
    .catch((e: Object) => {
      dispatch(failed(e.message));
    });
  };
}

export function doSignIn(username: string, email: string, password: string, image): Function {
  return (dispatch: Function): Promise<*> => {
    dispatch(loginRequested());
    const body = new FormData();
    body.append('username', username);
    body.append('password', password);
    body.append('email', email);
    
    if (image) {
      body.append('picture', {
        uri: image.uri,
        type: image.type,
        name: image.fileName,
      });
    }
    return fetch(`${config.API_URL}/users/register`, {
      method: 'POST',
      headers: {
        Accept: 'application/json',
      },
      body,
    })
    .then((response: Object): Object => response.json())
    .then(({token}): Promise => {
      dispatch(success(username, token));
      return AsyncStorage.setItem('agorask:user', JSON.stringify({username, token}));
    })
    .then(() => {
    })
    .catch(e => {
      dispatch(failed(e.message));
    });
  };
}
