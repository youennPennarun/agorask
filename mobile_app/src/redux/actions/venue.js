/* @flow */
/* global fetch */

export const FETCHING_LIST = 'FETCHING_LIST_VENUES';
export const SUCCESS_LIST = 'SUCCESS_LIST_VENUES';
export const FAILURE_LIST = 'FAILUR_LISTE_VENUES';

const API_URL = 'http://192.168.0.10:3000/venues';

export function getVenuesWithTasksNearPosition(location: {latitude: number, longitude: number}, radius = 0.5) {
  console.log(location);
  return dispatch => {
    dispatch({
      type: FETCHING_LIST,
    });
    fetch(API_URL, {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        lat: location.latitude,
        lng: location.longitude,
        radius,
        onlyWithTasks: true,
      }),
    })
    .then((response): Promise<*> => response.json())
    .then((venues: Array) => {
      dispatch({
        type: SUCCESS_LIST,
        venues,
      });
    })
    .catch(error => {
      dispatch({
        type: FAILURE_LIST,
        error,
      });
    });
  };
}
