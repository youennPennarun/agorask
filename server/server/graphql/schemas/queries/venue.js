const {
  GraphQLList,
  GraphQLID,
  GraphQLObjectType,
  GraphQLNonNull,
  GraphQLFloat,
  GraphQLString,
} = require('graphql');

const VenueType = require('../../types/venues');
const {getFields} = require('../../../utils/getGraphQLSelection');

const VenueService = require('../../../services/Venue');

const co = require('co');

const venuesWithinRadius = {
  type: new GraphQLList(VenueType),
  args: {
    lat: {
      type: new GraphQLNonNull(GraphQLFloat),
    },
    lng: {
      type: new GraphQLNonNull(GraphQLFloat),
    },
    radius: {
      name: 'radius',
      type: GraphQLFloat,
      description: 'Limit results to venues within this many meters.',
    },
  },
  resolve: function (root, {lat, lng, radius}, options, fields) {
    const fieldsNames = getFields(fields, Object);
    return co(VenueService.getVenuesWithinRadiusWithTasks({lat, lng}, radius, fieldsNames));
  },
};

const searchVenues = {
  type: new GraphQLList(VenueType),
  args: {
    lat: {
      type: new GraphQLNonNull(GraphQLFloat),
    },
    lng: {
      type: new GraphQLNonNull(GraphQLFloat),
    },
    radius: {
      name: 'radius',
      type: GraphQLFloat,
      description: 'Limit results to venues within this many meters.',
    },
    query: {
      type: new GraphQLNonNull(GraphQLString)
    },
  },
  resolve: function (root, {lat, lng, radius, query}, options) {
    return co(VenueService.searchVenue({lat, lng}, query, radius));
  },
};

const venue = {
  type: VenueType,
  args: {
    id: {
      type: new GraphQLNonNull(GraphQLID),
    },
    source: {
      type: GraphQLString,
    },
  },
  resolve: function (root, {id, source}, options, info) {
    const fieldsNames = getFields(info, Object);
    return co(VenueService.getVenue(id, source, fieldsNames));
  },
};

module.exports = {
  venue,
  searchVenues,
  venuesWithinRadius,
};
