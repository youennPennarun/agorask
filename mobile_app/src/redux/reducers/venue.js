import {FETCHING_LIST, SUCCESS_LIST, FAILURE_LIST, FETCHING_VENUE, UPDATE_SELECTED_VENUE, PUSH} from '../actions/venue';
import {FETCHING_TASK, SUCCESS_FETCH_TASK} from '../actions/task';


type VenueStateType = {
  version: Number,
  isFetching: Boolean,
  error: ?Object,
  venues: Array<Object>,
};
export const venuesDefaultState: VenueStateType = {
  version: 0,
  isFetching: false,
  error: null,
  venues: [],
};

export const selectedVenueDefaultState = {
  isFetching: false,
  _id: null,
  source: null,
  name: '',
  venue: {
    tasks: [],
  },
};

export const defaultTaskState = {
  isFetching: false,
  error: null,
  _id: null,
  title: '',
  answers: [],
};

function task(state = defaultTaskState, action) {
  switch (action.type) {
    case FETCHING_TASK:
      if (action.id === state._id) {
        return {
          ...state,
          isFetching: true,
          error: null,
        };
      }
      return state;
    case SUCCESS_FETCH_TASK:
      if (action.id === state._id) {
        return {
          ...state,
          ...action.task,
          isFetching: false,
          error: null,
        };
      }
      return state;
    default:
      return state;
  }
}

function selectedVenue(state = selectedVenueDefaultState, action) {
  switch (action.type) {
    case FETCHING_VENUE:
      return {
      ...state,
      isFetching: true,
    };
    case UPDATE_SELECTED_VENUE:
      return {
        ...state,
        venue: action.venue,
        isFetching: false,
      };
    case FETCHING_TASK:
      return {
        ...state,
        venue: {
          ...state.venue,
          tasks: state.venue.tasks.map(taskState => task(taskState, action)),
        },
      };
    case SUCCESS_FETCH_TASK:
      return {
        ...state,
        venue: {
          ...state.venue,
          tasks: state.venue.tasks.map(taskState => task(taskState, action)),
        },
      };
  default:
    return state;
  }
}

function venues(state: VenueStateType = venuesDefaultState, action): VenueStateType {
  switch (action.type) {
    case FETCHING_LIST:
      return {
        ...state,
        isFetching: true,
        error: null,
      };
    case FAILURE_LIST:
      return {
        ...state,
        isFetching: false,
        error: action.error,
      };
    case SUCCESS_LIST:
      return {
        ...state,
        version: state.version + 1,
        isFetching: false,
        error: null,
        venues: action.venues,
      };
    case PUSH: {
      const existingVenueIndex = state.venues.findIndex(venue => 
                    (venue._id === action.venue._id || venue.foursquareId === action.venue.foursquareId));
      let newVenues = [];
      if (existingVenueIndex > -1) {
        newVenues = [
          ...state.venues.slice(0, existingVenueIndex),
          action.venue,
          ...state.venues.slice(existingVenueIndex + 1),
        ];
      } else {
        newVenues = [
          ...state.venues,
          action.venue,
        ];
      }
      return {
        ...state,
        venues: newVenues,
      };
    }
    default:
      return state;
  }
}

export default {
  venues,
  selectedVenue,
};
