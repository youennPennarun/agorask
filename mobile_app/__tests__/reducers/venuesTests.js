/* global jest, expect, describe, it */

import 'react-native';
import React from 'react'; // eslint-ignore-line

import VenueReducers, {venuesDefaultState} from '../../src/redux/reducers/venue';
import {FETCHING_LIST, SUCCESS_LIST, FAILURE_LIST} from '../../src/redux/actions/venue';

it('default venues state', () => {
  const newState = VenueReducers.venues(undefined, {});
  expect(newState).toMatchSnapshot();
});

it('fetching venues', () => {
  const newState = VenueReducers.venues(venuesDefaultState, {
    type: FETCHING_LIST,
  });
  expect(newState).toMatchSnapshot();
});

it('fetch venues successful', () => {
  const newState = VenueReducers.venues(venuesDefaultState, {
    type: SUCCESS_LIST,
    venues: [{_id: 'id', name: 'venueTest'}],
  });
  expect(newState).toMatchSnapshot();
});
it('fetch venues fail', () => {
  const error = new Error('Unable to fetch venues');
  const newState = VenueReducers.venues(venuesDefaultState, {
    type: FAILURE_LIST,
    error,
  });
  expect(newState).toMatchSnapshot();
});


