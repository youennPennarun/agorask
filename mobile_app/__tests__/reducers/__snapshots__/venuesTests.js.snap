exports[`test renders correctly 1`] = `
Object {
  "error": null,
  "isFetching": false,
  "venues": Array [
    Object {
      "_id": "id",
      "name": "venueTest",
    },
  ],
  "version": 1,
}
`;
